### DISCUSSION TITLE
Log Event not published

### DISCUSSION THREAD

**Discussion by Soliton321** [jongpie/NebulaLogger#234] [Q&A] [answered]
hey there,

first of all great thanks for your contribution. I work as developer/architect in our company. Currently we are in final UAT introduing nebula as enterprise API logger. Its really a time+resource saver! 

we installed logger as managed package in our full sandbox. We exposed an REST-Webservice for receiving data from our consumer. REST Webservice should simply receive payload and forward it to nebula logger. We use default savemethod Event_BUS. However we discovered during UAT that a few logger platform events invoked by client are simply not published.  The request payload and statusCode were both fine. We can even replay the request sent afterwards without any issue. Currently out of 100+ requests 1 log is missing.  

From what I understand so far on documenation is that Salesforce PE itself is not as reliable as database transaction, such as DML or REST API. You can find it at bottom of documentation https://developer.salesforce.com/docs/atlas.en-us.platform_events.meta/platform_events/platform_events_considerations.htm 

Is there a way to intercept and recover these missing event in logger? Would the Publish Behavior "Publish after Commit" on PE help us?

By the way, I tried to switch to savemethod REST in custom setting, however no log entry is created on log. Did I miss something?

thanks again in advance

cheers

**Comment by jongpie** [comment, marked as answer]
Hi @Soliton321 - glad to hear that you're considering using Nebula Logger! One piece of feedback - I would strongly suggest that you consider using the unlocked package instead of the managed package. The unlocked package provides [a lot more functionality that is not possible with the managed package](https://github.com/jongpie/NebulaLogger/wiki/Unlocked-vs-Managed-Package).

Regarding the publish behavior "Publish after Commit", that would not be a useful change - with that behavior, any errors/exceptions will prevent the platform events from publishing, which means you would not be able to log anything when an exception occurs in your org (and in my opinion, logging exceptions is probably the most important time to be able to log). 

As far as your issue with log entries not saving - are you able to reproduce the issue consistently? Does it only happen in some scenarios? It's possible that there is a bug within Nebula Logger that's preventing the log entries from being saved - if you have any more details (or code) you can provide, that would be helpful in trying to determine the root cause of the issue. You can try temporarily using the save method `SYNCHRONOUS_DML` - this save method skips publishing the platform events, and instead immediately inserts `Log__c` and `LogEntry__c` records. It will also help surface any internal errors that might be occurring within Nebula Logger - if the issue is related to a bug within Nebula Logger, then this save method will cause an error to be thrown during your Apex transaction. Assuming you can consistently reproduce the issue, then try this option & let me know if it provides any more info on why your log entries aren't saving.

**Reply by Soliton321** [reply]
hey @jongpie thanks for suggestion. That helped me understand the logger more. We will take it into consideration to use unlocked package. 

I've seen you have made error logging for platform even https://github.com/jongpie/NebulaLogger/commit/ab7dc2064c756572c567e62e94e3bc799e1dcb7f Thank you for the effort. Could we expect to retrieve the `List<Database.SaveResult> saveResults` back in caller code in a future release?  Since then it would be possible for user to perform their own error handling.

Regarding the error "log entries not saved using `REST `savemode". I was able to reproduce the error. This occurred when ever I invoke `parseStackTrace`. The error somehow disappeared, after I renamed my apex method name for `REST POST` where nebula logger is invoked.  Otherwise error was silenced. Consumer of the exposed Apex REST webservice received success message. I even tried to made another insert in the next line as code below. Log will be created without logEntry, some fields like startTime, total CPU will be left empty. However, the account in the next line will be inserted. 
  
` Nebula.logger.info(payloadStr).parseStackTrace(new DmlException().getStackTraceString());`
` Nebula.Logger.saveLog();`
` insert Account(name='test');`

Using `parseStackTrace ` combined with `SYNCHRONOUS_DML`, I received error in `You have uncommitted work pending. Please commit or rollback before calling out`. Once again, this error disappeared after I renamed the method name where logging is invoked.


Hope this could help improve the error handling for the future.

**Reply by jongpie** [reply]
@Soliton321 thanks for the feedback! I like the idea of having the `saveLog()` method return the save results - I've added issue #242 to get this added in a future release!
